File: tests/e2e/go_client/routines_scripting_test.go
// Phase 6 routines + scripting E2E for the bqemulator via the Google
// Go BigQuery client. Matches the Phase 6 ship criterion.

package e2e

import (
	"context"
	"fmt"
	"testing"
	"time"

	"cloud.google.com/go/bigquery"
	"google.golang.org/api/iterator"
	"google.golang.org/api/option"
)

func routines_scriptingClient(ctx context.Context, t *testing.T) *bigquery.Client {
	t.Helper()
	client, err := bigquery.NewClient(
		ctx,
		"e2e-go-routines_scripting",
		option.WithEndpoint(bqAPIBase()),
		option.WithoutAuthentication(),
	)
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}
	return client
}

func createRoutine(ctx context.Context, t *testing.T, client *bigquery.Client, datasetID, routineID string, meta *bigquery.RoutineMetadata) {
	t.Helper()
	r := client.Dataset(datasetID).Routine(routineID)
	if err := r.Create(ctx, meta); err != nil {
		t.Fatalf("create routine %s: %v", routineID, err)
	}
}

func TestRoutinesScriptingShipCriterion(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
	defer cancel()
	client := routines_scriptingClient(ctx, t)
	defer client.Close()

	dsID := "routines_scripting_go_ds"
	ds := client.Dataset(dsID)
	_ = ds.DeleteWithContents(ctx)
	if err := ds.Create(ctx, &bigquery.DatasetMetadata{Location: "US"}); err != nil {
		t.Fatalf("create dataset: %v", err)
	}
	defer func() {
		if err := ds.DeleteWithContents(ctx); err != nil {
			t.Logf("cleanup: %v", err)
		}
	}()

	createRoutine(ctx, t, client, dsID, "sql_inc", &bigquery.RoutineMetadata{
		Type:     "SCALAR_FUNCTION",
		Language: "SQL",
		Arguments: []*bigquery.RoutineArgument{
			{Name: "x", DataType: &bigquery.StandardSQLDataType{TypeKind: "INT64"}},
		},
		ReturnType: &bigquery.StandardSQLDataType{TypeKind: "INT64"},
		Body:       "x + 1",
	})

	createRoutine(ctx, t, client, dsID, "js_double", &bigquery.RoutineMetadata{
		Type:     "SCALAR_FUNCTION",
		Language: "JAVASCRIPT",
		Arguments: []*bigquery.RoutineArgument{
			{Name: "x", DataType: &bigquery.StandardSQLDataType{TypeKind: "INT64"}},
		},
		ReturnType: &bigquery.StandardSQLDataType{TypeKind: "INT64"},
		Body:       "return x * 2;",
	})

	createRoutine(ctx, t, client, dsID, "one_to_n", &bigquery.RoutineMetadata{
		Type:     "TABLE_VALUED_FUNCTION",
		Language: "SQL",
		Arguments: []*bigquery.RoutineArgument{
			{Name: "n", DataType: &bigquery.StandardSQLDataType{TypeKind: "INT64"}},
		},
		Body: "SELECT i AS value FROM UNNEST(GENERATE_ARRAY(1, n)) AS i",
	})

	script := fmt.Sprintf(`
DECLARE n INT64 DEFAULT 3;
DECLARE total INT64 DEFAULT 0;
BEGIN
  FOR row IN (SELECT value FROM %s.one_to_n(n)) DO
    SET total = total + %s.js_double(%s.sql_inc(row.value));
  END FOR;
EXCEPTION WHEN ERROR THEN
  SET total = -1;
END;
IF total > 0 THEN
  SELECT total AS answer;
ELSE
  SELECT -1 AS answer;
END IF;
`, dsID, dsID, dsID)

	q := client.Query(script)
	it, err := q.Read(ctx)
	if err != nil {
		t.Fatalf("query: %v", err)
	}
	var row []bigquery.Value
	if err := it.Next(&row); err != nil && err != iterator.Done {
		t.Fatalf("row: %v", err)
	}
	if v, ok := row[0].(int64); !ok || v != 18 {
		t.Fatalf("expected answer=18, got %v", row[0])
	}
}

// TestScriptedCreateSchemaIsListed checks that a CREATE SCHEMA inside a
// multi-statement script registers the dataset in the catalog so it
// surfaces via datasets.list and datasets.get. A single-statement
// CREATE SCHEMA takes the executor fast path; the trailing SELECT tips
// this job into the scripting interpreter, whose DDL-sync hook must run.
func TestScriptedCreateSchemaIsListed(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
	defer cancel()
	client := routines_scriptingClient(ctx, t)
	defer client.Close()

	dsID := "scripted_created_schema_go_ds"
	ds := client.Dataset(dsID)
	// Guard against a stale dataset left by an interrupted run.
	_ = ds.DeleteWithContents(ctx)
	defer func() {
		if err := ds.DeleteWithContents(ctx); err != nil {
			t.Logf("cleanup: %v", err)
		}
	}()

	script := fmt.Sprintf("CREATE SCHEMA `%s`;\nSELECT 1 AS n;", dsID)
	if _, err := client.Query(script).Read(ctx); err != nil {
		t.Fatalf("script query: %v", err)
	}

	found := false
	it := client.Datasets(ctx)
	for {
		d, err := it.Next()
		if err == iterator.Done {
			break
		}
		if err != nil {
			t.Fatalf("list datasets: %v", err)
		}
		if d.DatasetID == dsID {
			found = true
			break
		}
	}
	if !found {
		t.Fatalf("dataset %q absent from datasets.list after scripted CREATE SCHEMA", dsID)
	}

	if _, err := ds.Metadata(ctx); err != nil {
		t.Fatalf("datasets.get %q: %v", dsID, err)
	}
}

// TestScriptEndingInDDLReturnsEmpty verifies last-statement-wins: a
// multi-statement script ending in DDL returns an empty result, not the
// prior SELECT's rows (BigQuery returns the final statement's result set).
func TestScriptEndingInDDLReturnsEmpty(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
	defer cancel()
	client := routines_scriptingClient(ctx, t)
	defer client.Close()

	dsID := "script_result_ddl_last_go"
	ds := client.Dataset(dsID)
	_ = ds.DeleteWithContents(ctx)
	if err := ds.Create(ctx, &bigquery.DatasetMetadata{Location: "US"}); err != nil {
		t.Fatalf("create dataset: %v", err)
	}
	defer func() {
		if err := ds.DeleteWithContents(ctx); err != nil {
			t.Logf("cleanup: %v", err)
		}
	}()

	script := fmt.Sprintf("SELECT 1 AS a;\nCREATE TABLE `%s.%s.trailing` (id INT64)", client.Project(), dsID)
	it, err := client.Query(script).Read(ctx)
	if err != nil {
		t.Fatalf("script query: %v", err)
	}
	var row []bigquery.Value
	if err := it.Next(&row); err != iterator.Done {
		t.Fatalf("expected empty result (iterator.Done), got row=%v err=%v", row, err)
	}
}

// TestRoutinesScriptingSingleRoutineDDLStatementType verifies that a
// single routine DDL statement reports BigQuery's statementType, and that
// DROP routine statements execute against the live container. CREATE
// PROCEDURE reports SCRIPT — BigQuery classifies a procedure definition
// as a script. Pinned by the routines_scripting/routine_ddl_*
// conformance corpus.
func TestRoutinesScriptingSingleRoutineDDLStatementType(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
	defer cancel()
	client := routines_scriptingClient(ctx, t)
	defer client.Close()

	dsID := "routine_ddl_go"
	ds := client.Dataset(dsID)
	_ = ds.DeleteWithContents(ctx)
	if err := ds.Create(ctx, &bigquery.DatasetMetadata{Location: "US"}); err != nil {
		t.Fatalf("create dataset: %v", err)
	}
	defer func() {
		if err := ds.DeleteWithContents(ctx); err != nil {
			t.Logf("cleanup: %v", err)
		}
	}()

	runDDL := func(sql string) string {
		job, err := client.Query(sql).Run(ctx)
		if err != nil {
			t.Fatalf("run %q: %v", sql, err)
		}
		status, err := job.Wait(ctx)
		if err != nil {
			t.Fatalf("wait %q: %v", sql, err)
		}
		if err := status.Err(); err != nil {
			t.Fatalf("job %q failed: %v", sql, err)
		}
		qs, ok := status.Statistics.Details.(*bigquery.QueryStatistics)
		if !ok {
			t.Fatalf("no QueryStatistics for %q", sql)
		}
		return qs.StatementType
	}

	proj := client.Project()
	cases := []struct{ sql, want string }{
		{fmt.Sprintf("CREATE FUNCTION `%s.%s.add_one`(x INT64) RETURNS INT64 AS (x + 1)", proj, dsID), "CREATE_FUNCTION"},
		{fmt.Sprintf("CREATE TABLE FUNCTION `%s.%s.one_row`(n INT64) AS SELECT n AS x", proj, dsID), "CREATE_TABLE_FUNCTION"},
		{fmt.Sprintf("CREATE PROCEDURE `%s.%s.noop`() BEGIN SELECT 1 AS one; END", proj, dsID), "SCRIPT"},
		{fmt.Sprintf("DROP FUNCTION `%s.%s.add_one`", proj, dsID), "DROP_FUNCTION"},
		{fmt.Sprintf("DROP PROCEDURE `%s.%s.noop`", proj, dsID), "DROP_PROCEDURE"},
		{fmt.Sprintf("DROP TABLE FUNCTION `%s.%s.one_row`", proj, dsID), "DROP_TABLE_FUNCTION"},
	}
	for _, c := range cases {
		if got := runDDL(c.sql); got != c.want {
			t.Fatalf("statementType for %q = %q, want %q", c.sql, got, c.want)
		}
	}
}

File: tests/e2e/nodejs_client/test/routines_scripting.test.js
/**
 * E2E: Phase 6 routines + scripting against the live container via the
 * @google-cloud/bigquery Node.js client.
 */

const { describe, it, before, after } = require("node:test");
const assert = require("node:assert/strict");

const REST_URL = process.env.BQEMU_REST_URL || "http://localhost:9050";
const PROJECT = "e2e-nodejs-routines_scripting";
const DATASET = "routines_scripting_node_ds";

function makeClient() {
  const { BigQuery } = require("@google-cloud/bigquery");
  const { OAuth2Client } = require("google-auth-library");

  const fake = new OAuth2Client();
  fake.credentials = { access_token: "anonymous" };

  return new BigQuery({
    projectId: PROJECT,
    apiEndpoint: REST_URL,
    authClient: fake,
    autoRetry: false,
  });
}

async function cleanup(client) {
  try {
    await client.dataset(DATASET).delete({ force: true });
  } catch (_) {
    /* ignore */
  }
}

async function createRoutine(client, body) {
  // The Node.js SDK's routines API uses request() under the hood via
  // dataset().routine(id).create(metadata). The metadata must match
  // the REST wire format exactly.
  const ds = client.dataset(DATASET);
  await ds.routine(body.routineReference.routineId).create(body);
}

describe("bqemulator Phase 6 routines + scripting (Node.js)", () => {
  let client;
  before(async () => {
    client = makeClient();
    await client.createDataset(DATASET, { location: "US" }).catch(() => {});
  });
  after(async () => {
    await cleanup(client);
  });

  it("registers SQL + JS + TVF and runs the ship-criterion script", async () => {
    const ref = (rid) => ({
      projectId: PROJECT,
      datasetId: DATASET,
      routineId: rid,
    });

    await createRoutine(client, {
      routineReference: ref("sql_inc"),
      routineType: "SCALAR_FUNCTION",
      language: "SQL",
      arguments: [{ name: "x", dataType: { typeKind: "INT64" } }],
      returnType: { typeKind: "INT64" },
      definitionBody: "x + 1",
    });

    await createRoutine(client, {
      routineReference: ref("js_double"),
      routineType: "SCALAR_FUNCTION",
      language: "JAVASCRIPT",
      arguments: [{ name: "x", dataType: { typeKind: "INT64" } }],
      returnType: { typeKind: "INT64" },
      definitionBody: "return x * 2;",
    });

    await createRoutine(client, {
      routineReference: ref("one_to_n"),
      routineType: "TABLE_VALUED_FUNCTION",
      language: "SQL",
      arguments: [{ name: "n", dataType: { typeKind: "INT64" } }],
      definitionBody:
        "SELECT i AS value FROM UNNEST(GENERATE_ARRAY(1, n)) AS i",
    });

    const script = `
DECLARE n INT64 DEFAULT 3;
DECLARE total INT64 DEFAULT 0;
BEGIN
  FOR row IN (SELECT value FROM ${DATASET}.one_to_n(n)) DO
    SET total = total + ${DATASET}.js_double(${DATASET}.sql_inc(row.value));
  END FOR;
EXCEPTION WHEN ERROR THEN
  SET total = -1;
END;
IF total > 0 THEN
  SELECT total AS answer;
ELSE
  SELECT -1 AS answer;
END IF;
`;
    const [rows] = await client.query(script);
    assert.equal(rows[0].answer, 18);
  });

  it("INFORMATION_SCHEMA.ROUTINES lists everything created", async () => {
    const sql = `SELECT routine_name FROM \`${PROJECT}\`.${DATASET}.INFORMATION_SCHEMA.ROUTINES ORDER BY routine_name`;
    const [rows] = await client.query(sql);
    const names = rows.map((r) => r.routine_name);
    assert.deepEqual(names, ["js_double", "one_to_n", "sql_inc"]);
  });

  it("registers a CREATE SCHEMA created inside a multi-statement script", async () => {
    // A single-statement CREATE SCHEMA takes the executor fast path; the
    // trailing SELECT tips this job into the scripting interpreter, whose
    // DDL-sync hook must register the dataset so it surfaces via
    // datasets.list and datasets.get.
    const scriptedDs = "scripted_created_schema_node_ds";
    // Guard against a stale dataset left by an interrupted run.
    await client
      .dataset(scriptedDs)
      .delete({ force: true })
      .catch(() => {});
    try {
      await client.query(`CREATE SCHEMA \`${scriptedDs}\`;\nSELECT 1 AS n;`);

      const [datasets] = await client.getDatasets();
      const ids = datasets.map((d) => d.id);
      assert.ok(
        ids.includes(scriptedDs),
        `dataset ${scriptedDs} absent from datasets.list after scripted CREATE SCHEMA`,
      );

      const [exists] = await client.dataset(scriptedDs).exists();
      assert.ok(exists, `datasets.get failed for ${scriptedDs}`);
    } finally {
      await client
        .dataset(scriptedDs)
        .delete({ force: true })
        .catch(() => {});
    }
  });

  it("returns an empty result for a script ending in DDL", async () => {
    // Last-statement-wins: a trailing DDL has no result set, so the prior
    // SELECT's rows must not leak into the script result.
    const ds = "script_result_ddl_last_node";
    await client
      .dataset(ds)
      .delete({ force: true })
      .catch(() => {});
    await client.createDataset(ds, { location: "US" });
    try {
      const script = `SELECT 1 AS a;\nCREATE TABLE \`${PROJECT}.${ds}.trailing\` (id INT64)`;
      const [rows] = await client.query(script);
      assert.equal(rows.length, 0);
    } finally {
      await client
        .dataset(ds)
        .delete({ force: true })
        .catch(() => {});
    }
  });
});
